casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala

```diff
@@ -258,43 +260,54 @@ object ProtoUtil {
    */
   def conflicts[F[_]: Monad: BlockStore](
       b1: BlockMessage,
       b2: BlockMessage,
       genesis: BlockMessage,
       dag: BlockDag
-  ): F[Boolean] =
+  ): F[Boolean] = {
+    val b1MetaData       = dag.dataLookup.get(b1.blockHash)
+    val b2MetaData       = dag.dataLookup.get(b2.blockHash)
+    val blockMetaDataSeq = Vector(b1MetaData, b2MetaData).flatten
+    if (blockMetaDataSeq.size == 2) {
+      implicit val ordering    = BlockDag.deriveOrdering(dag)
+      val uncommonAncestorsMap = DagOperations.uncommonAncestors(blockMetaDataSeq, dag.dataLookup)
+      val (b1AncestorsMap, b2AncestorsMap) = uncommonAncestorsMap.partition {
+        case (_, bitSet) => bitSet == BitSet(0)
+      }
+      val b1AncestorsMeta = b1AncestorsMap.keys
+      val b2AncestorsMeta = b2AncestorsMap.keys
+      for {
+        b1AncestorChannels <- buildBlockAncestorChannels[F](b1AncestorsMeta.toList)
+        b2AncestorChannels <- buildBlockAncestorChannels[F](b2AncestorsMeta.toList)
+      } yield b1AncestorChannels.intersect(b2AncestorChannels).nonEmpty
+    } else {
+      false.pure[F]
+    }
+  }
+
+  private def buildBlockAncestorChannels[F[_]: Monad: BlockStore](
+      blockAncestorsMeta: List[BlockMetadata]
+  ): F[Set[BlockHash]] =
     for {
-      gca <- DagOperations.greatestCommonAncestorF[F](b1, b2, genesis, dag)
-      result <- if (gca == b1 || gca == b2) {
-                 //blocks which already exist in each other's chains do not conflict
-                 false.pure[F]
-               } else {
-                 def getDeploys(b: BlockMessage) =
-                   for {
-                     bAncestors <- DagOperations
-                                    .bfTraverseF[F, BlockMessage](List(b))(
-                                      ProtoUtil.unsafeGetParents[F]
-                                    )
-                                    .takeWhile(_ != gca)
-                                    .toList
-                     deploys = bAncestors
-                       .flatMap(b => {
-                         b.body.map(_.deploys.flatMap(_.deploy)).getOrElse(List.empty[Deploy])
-                       })
-                       .toSet
-                   } yield deploys
-                 for {
-                   b1Deploys <- getDeploys(b1)
-                   b2Deploys <- getDeploys(b2)
-                 } yield b1Deploys.intersect(b2Deploys).nonEmpty
-               }
-    } yield result
+      maybeAncestors <- blockAncestorsMeta.traverse(
+                         blockAncestorMeta => BlockStore[F].get(blockAncestorMeta.blockHash)
+                       )
+      ancestors      = maybeAncestors.flatten
+      ancestorEvents = ancestors.flatMap(_.getBody.deploys.flatMap(_.log))
+      ancestorChannels = ancestorEvents.flatMap {
+        case Event(Produce(produce: ProduceEvent)) =>
+          Set(produce.channelsHash)
+        case Event(Consume(consume: ConsumeEvent)) =>
+          consume.channelsHashes.toSet
```

Comment:
After seeing what you are doing here I guess it would make sense to change the type of the `channelsHashes` container to `Set` in `Consume`. Could be done in a separate PR

Reply:
We need them ordered in the proto though because these Events all get hashed and the block hash needs to be deterministic.